### XMAC02 - Prova Prática 2

In [1]:
import numpy as np
import pandas as pd
from scipy.stats import norm
import scipy.stats as stats
from statsmodels.stats import weightstats, proportion
from scipy.stats import f

/home/aluno/anaconda3/lib/python3.11/site-packages/pandas/core/arrays/masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).
  from pandas.core import (


### Questão 1
Uma empresa fabricante de pneus automotivos descobriu que a vida útil dos pneus segue uma distribuição normal com média de 40.000 km e desvio padrão de 3.000 km. Acredita-se que uma mudança no processo de produção resultará em um produto melhor e, por isso, a empresa desenvolveu um novo pneu. Uma amostra de 100 pneus novos foi selecionada. A empresa constatou que a vida útil média desses novos pneus é de 40.900 km. Pode-se concluir que o novo pneu é significativamente melhor que o antigo, usando um nível de significância de 0,01?

Testar se a média nova (uNova) é maior que a média antiga (uAntiga) da vida útil dos pneus, utilizando o teste Z.

Z calc = (Média observada - média populacional) / (desvio padrão / raiz de n) 

H0: uNova <= uAntiga \
Ha: uNova > uAntiga \
alpha = 0,01

In [25]:
z_calc = (40900 - 40000) / (3000 / np.sqrt(100))
z_calc

3.0

In [27]:
norm.isf(0.01) # z_crit 1 cauda a direita

2.3263478740408408

Como z_calc (3.0) é maior que z_crit (2.36), rejeitamos a H0. Portanto, podemos afirmar que o novo pneu é significativamente melhor que o antigo, tendo em vista que sua média de vida últil é maior.

### Questão 2
Sabe-se que a alcalinidade, em miligramas por litro, da água nos trechos superiores dos rios de uma determinada região segue uma distribuição normal com um desvio padrão de 10 mg/l. As leituras de alcalinidade nos trechos inferiores dos rios da mesma região também seguem uma distribuição normal, mas com um desvio padrão de 25 mg/l. Dez leituras de alcalinidade são feitas nos trechos superiores de um rio da região e quinze nos trechos inferiores do mesmo rio, com os seguintes resultados.

trechos_sup = [91, 75, 91, 88, 94, 63, 86, 77, 71, 69] \
trechos_inf = [86, 95, 135, 121, 68, 64, 113, 108, 79, 62, 143, 108, 121, 85, 97]

Investigue, com um nível de confiança de 95%, a afirmação de que a verdadeira alcalinidade média da água no curso inferior deste rio é maior do que no curso superior.

Testar se a alcalinidade média dos trechos inferiores (uTrechos_inf) é maior que nos trechos superiores (uTrechos_sup), usando Teste T pois `n < 20`, e independente, pois um resultado não afeta o outro. \
H0: uTrechos_inf <=  uTrechos_sup \
HA: uTrechos_inf >  uTrechos_sup \
alpha = 0.05

In [ ]:
trechos_inf = [86, 95, 135, 121, 68, 64, 113, 108, 79, 62, 143, 108, 121, 85, 97]
trechos_sup = [91, 75, 91, 88, 94, 63, 86, 77, 71, 69]

stats.ttest_ind(trechos_inf, trechos_sup, alternative='greater')

Ttest_indResult(statistic=2.177781517785407, pvalue=0.019965716484029193)

Como pValue é menor que 0.05, rejeitamos a H0. Portanto, a alcalinidade média dos trechos inferiores é MAIOR que a média dos trechos superiores.

### Questão 3
Abóboras foram cultivadas sob duas condições experimentais, x e y. Do experimento x, uma amostra de 11 abóboras foi coletada e o desvio padrão amostral observado no peso das abóboras foi 0.8. Do experimento y, uma amostra de 9 abóboras foi coletada e o desvio padrão amostral observado foi 0.5. Assumindo que as distribuições de peso são normais, teste a hipótese de que as variâncias verdadeiras são iguais, contra a alternativa de que não são, ao nível de 90% de confiança.

x: n = 11, s = 0.8 \
y: n = 9, s = 0.5

Teste de variância para verificar se a variância da amostra X (pX) é igual a variância da amostra Y (pY). \
H0: pX - pY = 0 \
HA: pX - pY <> 0 \
alpha = 0.1

In [7]:
fCal = (0.8 ** 2) / (0.5 ** 2)
fCal

2.5600000000000005

In [8]:
f.isf(0.05, 8, 10) # ponto critico à direita

3.071658385279037

In [9]:
f.isf(0.95, 8, 10) # ponto critico à esquerda


0.29876046194309663

Como `fCal` (2.56) está entre os pontos críticos, falhamos em rejeitar a H0. Portanto, a variância das amostras é suficientemente igual.

### Questão 4
Queremos testar a afirmação de que ouvir música clássica enquanto estuda faz com que os alunos concluam os trabalhos mais rapidamente. Pedimos a 10 alunos estudem em silêncio no primeiro semestre e estudem ouvindo música clássica no segundo semestre, e registramos o número médio de horas gasto executando trabalhos por semana em cada semestre. Os tempos gastos na execução de trabalhos em 10 semanas de estudo estão listados a seguir. Podemos concluir que estudar ouvindo música clássica reduz o número de horas gastas por semana na execução dos trabalhos escolares? Analise o valor de t_calc (statistic) com confiança de 95%.

Importante: Sua conclusão não deve utilizar o pvalue. Conclua utilizando o valor de t_calc e o t_crítico obtido da tabela t. Note que não é necessário fazer o cálculo à mão, já que o t_calc (statistic) é fornecido pelo Python.

Em silêncio: 14, 13, 16, 21, 15, 19, 11, 20, 19, 16 \
Ouvindo música: 12, 13, 15, 22, 16, 19, 8, 17, 18, 17

Testar se a média de horas estudadas em silêncio (pSilêncio) é maior que ouvindo música (pMusica). \
H0: pSilencio <= pMusica \
HA: pSilencio > pMusica \
alpha = 0.05 \
Obter o t_crítico utilizando isf. 

In [11]:
silencio = [14, 13, 16, 21, 15, 19, 11, 20, 19, 16]
musica = [12, 13, 15, 22, 16, 19, 8, 17, 18, 17]

stats.ttest_rel(silencio, musica, alternative='greater')

TtestResult(statistic=1.412612867392256, pvalue=0.09570152734951828, df=9)

In [23]:
norm.isf(0.05) # t_crit de 1 cauda

1.6448536269514729

Como t_calc (1.41) é menor que o valor de t_crit (1.64), falhamos em rejeitar a H0. Portanto, estudar ouvindo música clássica NÃO resulta em uma média menor de horas gastas. \
Isso é reforçado pelo valor de pValue ser maior que 0.05.

### Questão 5
O proprietário de um cinema quer saber se há alguma diferença no número de meninos e meninas que assistiram a um novo filme na semana passada. Para tanto, foram amostrados aleatoriamente 76 meninos e 75 meninas e descobriram que 45 meninos e 58 meninas assistiram ao filme. O que o proprietário pode concluir sobre a diferença de proporções com um nível de confiança de 99%?

Confirme o resultado do pvalue confrontando com o resultado do t_calc (statistic) comparado com o t_critico obtido da tabela Z. Só então decida se você falha em rejeitar H0 ou rejeita H0.

Fazer um teste de proporção para verificar se proporção de meninos que assistiram o filme é igual à proporção de meninas. \
p1 = qtd de meninos que assistiram / qtd total de meninos \
p2 = qtd de meninas que assistiram / qtd total de meninas \
H0: p1 - p2 = 0 \
HA: p1 - p2 <> 0 \
alpha = 0.01

In [18]:
proportion.test_proportions_2indep(45, 76, 58, 75)

<class 'statsmodels.stats.base.HolderTuple'>
statistic = -2.3955021617126655
pvalue = 0.0165976163678317
compare = 'diff'
method = 'agresti-caffo'
diff = -0.18122807017543863
ratio = 0.7656533575317603
odds_ratio = 0.4254727474972191
variance = 0.00542810125834494
alternative = 'two-sided'
value = 0
tuple = (-2.3955021617126655, 0.0165976163678317)

In [21]:
norm.ppf(0.005) # z_crit esquerda

-2.575829303548901

In [22]:
norm.isf(0.005) # z_crit direita


2.575829303548901

Como o valor de t_calc (-2.39) está entre os valores de z_crit, falhamos em rejeitar a H0 (o que também é reforçado por pValue ser maior que 0.01). Portanto, a proporção de meninos e meninas que assistiram o filme é suficientemente igual, apesar de um z_calc próximo do z_crit à esquerda, e um pValue muito próximo de alpha.

### Questão 6
Uma turma do quarto ano está estudando o meio ambiente. Uma das tarefas é cultivar feijões em diferentes tipos de solo. Pedro escolheu cultivar seus feijões em terra encontrada do lado de fora da sala de aula, misturada com fiapos de secadora. Ana escolheu cultivar seus feijões em terra para vasos comprada em um viveiro local. Ivo escolheu cultivar seus feijões em terra do jardim de sua mãe. Nenhum produto químico foi usado nas plantas, apenas água. Elas foram cultivadas dentro da sala de aula, perto de uma grande janela. Cada criança cultivou cinco plantas. Ao final do período de crescimento, cada planta foi medida, registrando-se os dados (em polegadas).

pedro = [24, 21, 23, 30, 23] \
ana = [25, 31, 23, 20, 28] \
ivo = [23, 27, 22, 30, 20]

Faça um teste de hipótese capaz de avaliar a se o tipo de solo fez alguma diferença no crescimento do feijão. Utilize confiança de 95%.

Testar se as três médias são iguais, ou se há alguma diferença entre alguma das médias, utilizando ANOVA. \
H0: uPedro == uAna == uIvo \
HA: Alguma das médias é diferente \
alpha = 0.05

In [3]:
pedro = [24, 21, 23, 30, 23]
ana = [25, 31, 23, 20, 28]
ivo = [23, 27, 22, 30, 20]

stats.levene(pedro, ana, ivo)

LeveneResult(statistic=0.2857142857142858, pvalue=0.7564490529115285)

Como pValue obtido (0.76) é maior que 0.05, falhamos em rejeitar a H0. Portanto, o tipo de solo não fez nenhuma diferença no crescimento do feijão, tendo em vista que as três médias são suficientemente iguais.

### Questão 7
Considere o seguinte sistema linear: 

4x - 2y + 3z =  1 \
 x + 3y - 4z = -7 \
3x +  y + 2z =  5 

Utilizando eliminação gaussiana, obtenha uma matriz triangular, resolva o sistema por substituição e informe a solução do sistema.

4x - 2y + 3z =  1 \
 x + 3y - 4z = -7 \
3x +  y + 2z =  5  


Extraindo os números: \
4 -2 3 1 \
1 3 -4 -7 \
3 1 2 5

L2 = 4L2 - L1: \
4 -2 3 1 \
0 14 -19 -29 \
3 1 2 5 

L3 = 4L3 - 3L1: \
4 -2 3 1 \
0 14 -19 -29 \
0 10 -1 17 

L3 = 10L2 - 14L3: \
4 -2 3 1 \
0 14 -19 -29 \
0 0 -176 52  

Utilizando L3:
-176z = 52 -> z = 52/-176 = -0.295 = z

Substituindo em L2: \
14y -19z = -29 \
14y - 5,61 = -29 \
y = (-29 + 5,61) / 14 = 2.47 = y

Substituindo em L1:
4x -2y 3z = 1 \
4x -2*2.47 +3*(-0.295) = 1 \
4x -4.94 -0.885 = 1 \
x = 1.45625

Dessa forma:
x = 1,45625 \
y = 2.47 \
z = -0,295


In [28]:
A = np.array([[4, -2, 3],
             [1, 3, -4],
             [3, 1, 2]])

b = np.array([1, -7, 5])

solution = np.linalg.solve(A, b)

print(f'Beta0 = {solution[0]:9,.2f}')
print(f'Beta1 = {solution[1]:9,.2f}')
print(f'Beta2 = {solution[2]:9,.2f}')


Beta0 =     -1.00
Beta1 =      2.00
Beta2 =      3.00
